File: bindings/typescript/src/durable.ts
import { spawn, type ChildProcessWithoutNullStreams } from 'node:child_process';
import { resolveHostBinary } from './binary.js';

/** Exact decimal strings, never JavaScript floating-point cost estimates or receipts. */
export type ResourceAmounts = Readonly<Record<string, string>>;
export interface DurableWorkIdentity {
  readonly runId: string;
  readonly evaluationId: string;
  readonly attempt: number;
  readonly leaseId: string;
}
export interface DurableWorkConfig {
  readonly directory: string;
  readonly runId: string;
  readonly compatibilityHash: string;
  readonly limits: ResourceAmounts;
  readonly maximumWorkItems?: number;
  readonly maximumDeliveriesPerWork?: number;
  readonly maximumStateBytes?: number;
  readonly maximumPayloadBytes?: number;
  readonly maximumWorkers?: number;
  readonly leaseDurationMs?: number;
  readonly hostPath?: string;
  /** Prefix arguments, for example the managed host DLL. --durable is appended automatically. */
  readonly hostArgs?: readonly string[];
  readonly requestTimeoutMs?: number;
}
export interface DurableJob {
  readonly evaluationId: string;
  readonly attempt: number;
  readonly canonicalGenomeId: string;
  readonly payload: string;
  readonly tags?: readonly string[];
  readonly minimumResources?: ResourceAmounts;
  readonly estimated: ResourceAmounts;
  readonly maximum: ResourceAmounts;
}
export interface DurableWorker {
  readonly workerId: string;
  readonly compatibilityHash: string;
  readonly tags?: readonly string[];
  readonly capacity?: ResourceAmounts;
  readonly maximumConcurrentWork?: number;
}
export interface DurableLease {
  readonly identity: DurableWorkIdentity;
  readonly workerId: string;
  readonly canonicalGenomeId: string;
  readonly payload: string;
  readonly deliveryNumber: number;
  readonly expiresAt: string;
}
export type ReceiptOutcome = 'completed' | 'failed' | 'rejected' | 'canceled';
export type CommitDisposition = 'accepted' | 'duplicate' | 'stale' | 'duplicate-stale' | 'unknown-lease' | 'budget-violation';
export type HeartbeatStatus = 'renewed' | 'expired' | 'canceled' | 'completed' | 'unknown-lease';
export interface DurableReceipt {
  readonly identity: DurableWorkIdentity;
  readonly workerId: string;
  readonly payload: string;
  readonly provenance: string;
  readonly actual: ResourceAmounts;
  readonly outcome: ReceiptOutcome;
}
export interface DurableResult extends Omit<DurableReceipt, 'workerId'> { readonly accepted: boolean; }
export interface DurableStatus {
  readonly runId: string;
  readonly compatibilityHash: string;
  readonly wasRecovered: boolean;
  readonly sourceSessionId: string | null;
  readonly supportsExactSearchContinuation: false;
  readonly searchContinuationGuarantee: string;
  readonly searchState: 'not-owned';
  readonly spent: ResourceAmounts;
  readonly reserved: ResourceAmounts;
  readonly admitted: string;
  readonly settled: string;
  readonly denied: string;
  readonly maximumViolated: boolean;
}
export interface DurableEvaluationPayload {
  readonly schema: 1;
  readonly genomePayload: string;
  readonly canonicalGenomeId: string;
  readonly evaluationId: string;
  readonly attempt: number;
  readonly rootSeed: string;
  readonly seedStream: string;
}

const MAX_FRAME_BYTES = 16 * 1024 * 1024;
const MAX_INT32 = 2_147_483_647;
const INT64_MAX = 9223372036854775807n;
const UINT64_MAX = 18446744073709551615n;
type ObjectValue = Record<string, unknown>;
function record(value: unknown): value is ObjectValue { return value !== null && typeof value === 'object' && !Array.isArray(value); }
function integer(value: unknown, max: number = MAX_INT32): value is number {
  return typeof value === 'number' && Number.isSafeInteger(value) && value > 0 && value <= max;
}
function unsigned(value: unknown, max: bigint = INT64_MAX): value is string {
  return typeof value === 'string' && /^(0|[1-9][0-9]{0,19})$/.test(value) && BigInt(value) <= max;
}
function amounts(value: unknown): value is ResourceAmounts {
  return record(value) && Object.keys(value).length <= 32 && Object.entries(value).every(([key, amount]) =>
    key.trim().length > 0 && key.length <= 64 && !/[\u0000-\u001f\u007f-\u009f]/.test(key) &&
    typeof amount === 'string' && /^(0|[1-9][0-9]{0,28})(\.[0-9]{0,27}[1-9])?$/.test(amount));
}
function identity(value: unknown): value is DurableWorkIdentity {
  return record(value) && typeof value.runId === 'string' && value.runId.length > 0 && value.runId.length <= 256 &&
    unsigned(value.evaluationId) && integer(value.attempt) && typeof value.leaseId === 'string' && /^[0-9a-f]{32}$/.test(value.leaseId);
}
function lease(value: unknown): value is DurableLease {
  return record(value) && identity(value.identity) && typeof value.workerId === 'string' && value.workerId.length > 0 &&
    typeof value.canonicalGenomeId === 'string' && typeof value.payload === 'string' && integer(value.deliveryNumber, 16) &&
    typeof value.expiresAt === 'string' && Number.isFinite(Date.parse(value.expiresAt));
}
const outcomes = ['completed', 'failed', 'rejected', 'canceled'] as const;
const dispositions = ['accepted', 'duplicate', 'stale', 'duplicate-stale', 'unknown-lease', 'budget-violation'] as const;
const heartbeats = ['renewed', 'expired', 'canceled', 'completed', 'unknown-lease'] as const;
function member<T extends string>(value: unknown, choices: readonly T[]): value is T {
  return typeof value === 'string' && choices.includes(value as T);
}
function result(value: unknown): value is DurableResult {
  return record(value) && identity(value.identity) && typeof value.payload === 'string' && typeof value.provenance === 'string' &&
    amounts(value.actual) && member(value.outcome, outcomes) && typeof value.accepted === 'boolean';
}
function status(value: ObjectValue): DurableStatus {
  if (typeof value.runId !== 'string' || typeof value.compatibilityHash !== 'string' || typeof value.wasRecovered !== 'boolean' ||
      !(value.sourceSessionId === null || (typeof value.sourceSessionId === 'string' && /^[0-9a-f]{32}$/.test(value.sourceSessionId))) ||
      value.supportsExactSearchContinuation !== false || value.searchState !== 'not-owned' ||
      typeof value.searchContinuationGuarantee !== 'string' || !value.searchContinuationGuarantee.includes('fork') ||
      !amounts(value.spent) || !amounts(value.reserved) || !unsigned(value.admitted) || !unsigned(value.settled) ||
      !unsigned(value.denied) || typeof value.maximumViolated !== 'boolean') throw new DurableWorkError('Invalid durable status response.');
  return value as unknown as DurableStatus;
}
function boolean(value: unknown): boolean {
  if (typeof value !== 'boolean') throw new DurableWorkError('Missing boolean response field.');
  return value;
}
function receipt(value: unknown): DurableResult | null {
  if (value === null || result(value)) return value;
  throw new DurableWorkError('Invalid durable receipt response.');
}
function ticket(value: DurableWorkIdentity): void {
  if (!identity(value)) throw new DurableWorkError('The full original durable identity is required; evaluationId is a decimal string.');
}
function cost(value: ResourceAmounts): void {
  if (!amounts(value)) throw new DurableWorkError('Resources require bounded canonical decimal-string amounts.');
}

/** Parses the engine bridge envelope without losing 64-bit seeds or evaluation identities. */
export function parseDurableEvaluationPayload(json: string): DurableEvaluationPayload {
  if (Buffer.byteLength(json, 'utf8') > 1024 * 1024) throw new DurableWorkError('Evaluation envelope exceeds one MiB.');
  const value: unknown = JSON.parse(json);
  if (!record(value) || value.schema !== 1 || typeof value.genomePayload !== 'string' ||
      typeof value.canonicalGenomeId !== 'string' || !value.canonicalGenomeId.trim() || !unsigned(value.evaluationId) ||
      !integer(value.attempt) || !unsigned(value.rootSeed, UINT64_MAX) || !unsigned(value.seedStream, UINT64_MAX)) {
    throw new DurableWorkError('Invalid versioned evaluation envelope.');
  }
  return value as unknown as DurableEvaluationPayload;
}
export class DurableWorkError extends Error {
  constructor(message: string) { super(message); this.name = 'DurableWorkError'; }
}

/** Trusted local durable-control/worker endpoint. Always close; never automatically rerun unacknowledged physical work. */
export class DurableWorkClient {
  readonly #child: ChildProcessWithoutNullStreams;
  readonly #timeout: number;
  readonly #runId: string;
  readonly #pending = new Map<number, { accept: (value: ObjectValue) => void; reject: (error: Error) => void; timer: NodeJS.Timeout }>();
  #nextId = 1;
  #buffer = '';
  #stderr = '';
  #fatal: Error | undefined;
  #closing: Promise<void> | undefined;

  private constructor(child: ChildProcessWithoutNullStreams, timeout: number, runId: string) {
    this.#child = child; this.#timeout = timeout; this.#runId = runId;
    child.stderr.setEncoding('utf8');
    child.stderr.on('data', (chunk: string) => { this.#stderr = (this.#stderr + chunk).slice(-4096); });
    child.stdout.setEncoding('utf8');
    child.stdout.on('data', (chunk: string) => {
      if (this.#fatal) return;
      this.#buffer += chunk;
      for (;;) {
        const index = this.#buffer.indexOf('\n');
        const frame = index < 0 ? this.#buffer : this.#buffer.slice(0, index);
        if (Buffer.byteLength(frame, 'utf8') > MAX_FRAME_BYTES) { this.#fail('Oversized durable response; reconcile the original store.'); return; }
        if (index < 0) return;
        this.#buffer = this.#buffer.slice(index + 1);
        try {
          const response: unknown = JSON.parse(frame);
          if (!record(response) || !integer(response.id, Number.MAX_SAFE_INTEGER) || response.protocol !== 1 || typeof response.ok !== 'boolean') {
            throw new DurableWorkError('Invalid durable response or protocol downgrade.');
          }
          const waiter = this.#pending.get(response.id);
          if (!waiter) throw new DurableWorkError('Uncorrelated durable response.');
          this.#pending.delete(response.id); clearTimeout(waiter.timer); waiter.accept(response);
        } catch (error) { this.#fail(String(error)); return; }
      }
    });
    child.on('error', (error) => this.#fail(`Durable host failed to start: ${error.message}`));
    child.stdin.on('error', (error) => this.#fail(`Durable pipe failed: ${error.message}; reconcile the original store.`));
    child.on('close', (code, signal) => this.#fail(`Durable host closed (${code ?? signal}); reconcile unacknowledged work. ${this.#stderr}`));
  }
  get pid(): number | undefined { return this.#child.pid; }

  static async open(config: DurableWorkConfig): Promise<DurableWorkClient> {
    const { hostPath, hostArgs, requestTimeoutMs = 120_000, ...wire } = config;
    if (!integer(requestTimeoutMs)) throw new DurableWorkError('requestTimeoutMs must fit a positive Int32 timer.');
    cost(config.limits);
    const child = spawn(hostPath ?? resolveHostBinary(), [...(hostArgs ?? []), '--durable'], { stdio: ['pipe', 'pipe', 'pipe'], windowsHide: true });
    const client = new DurableWorkClient(child, requestTimeoutMs, config.runId);
    try {
      const opened = await client.#call('open', { config: wire }, status);
      if (opened.runId !== config.runId || opened.compatibilityHash !== config.compatibilityHash) throw new DurableWorkError('Durable host changed the requested run or compatibility.');
      return client;
    } catch (error) { client.#fail(String(error)); throw error; }
  }
  async enqueue(job: DurableJob): Promise<boolean> {
    if (!unsigned(job.evaluationId) || !integer(job.attempt)) throw new DurableWorkError('Invalid logical work identity.');
    cost(job.estimated); cost(job.maximum); if (job.minimumResources !== undefined) cost(job.minimumResources);
    return this.#call('enqueue', { job }, value => boolean(value.enqueued));
  }
  /** Null means no matching work now; it never signals completion or permits a budget reset. */
  async claim(worker: DurableWorker): Promise<DurableLease | null> {
    if (worker.capacity !== undefined) cost(worker.capacity);
    return this.#call('claim', { worker }, value => {
      if (value.available === false && value.lease === null) return null;
      if (value.available === true && lease(value.lease) && value.lease.workerId === worker.workerId && value.lease.identity.runId === this.#runId) return value.lease;
      throw new DurableWorkError('Invalid claim response.');
    });
  }
  async heartbeat(identity: DurableWorkIdentity, workerId: string): Promise<HeartbeatStatus> {
    ticket(identity);
    return this.#call('heartbeat', { identity, workerId }, value => {
      if (!member(value.status, heartbeats)) throw new DurableWorkError('Invalid heartbeat response.'); return value.status;
    });
  }
  async cancel(evaluationId: string, attempt: number): Promise<boolean> {
    if (!unsigned(evaluationId) || !integer(attempt)) throw new DurableWorkError('Invalid logical work identity.');
    return this.#call('cancel', { evaluationId, attempt }, value => boolean(value.canceled));
  }
  async commit(receipt: DurableReceipt): Promise<CommitDisposition> {
    ticket(receipt.identity); cost(receipt.actual);
    if (!member(receipt.outcome, outcomes)) throw new DurableWorkError('An explicit actual receipt outcome is required.');
    return this.#call('commit', { ...receipt }, value => {
      if (!member(value.disposition, dispositions)) throw new DurableWorkError('Invalid commit response.'); return value.disposition;
    });
  }
  async result(evaluationId: string, attempt: number): Promise<DurableResult | null> {
    if (!unsigned(evaluationId) || !integer(attempt)) throw new DurableWorkError('Invalid logical work identity.');
    return this.#call('result', { evaluationId, attempt }, value => {
      const found = receipt(value.result);
      if (found && (found.identity.runId !== this.#runId || found.identity.evaluationId !== evaluationId || found.identity.attempt !== attempt)) {
        throw new DurableWorkError('Result belongs to different logical work.');
      }
      return found;
    });
  }
  async delivery(identity: DurableWorkIdentity, workerId: string): Promise<DurableResult | null> {
    ticket(identity); return this.#call('delivery', { identity, workerId }, value => {
      const found = receipt(value.result);
      if (found && Object.entries(identity).some(([key, field]) => found.identity[key as keyof DurableWorkIdentity] !== field)) {
        throw new DurableWorkError('Receipt belongs to a different delivery.');
      }
      return found;
    });
  }
  /** Single-lease keyset page for reconciliation only, never permission to repeat physical execution. */
  async unsettled(workerId: string, afterLeaseId?: string): Promise<{ lease: DurableLease | null; nextAfterLeaseId: string | null }> {
    return this.#call('unsettled', { workerId, afterLeaseId }, value => {
      if (value.reconciliationOnly !== true || !(value.lease === null || (lease(value.lease) && value.lease.workerId === workerId && value.lease.identity.runId === this.#runId)) ||
          value.nextAfterLeaseId !== (value.lease?.identity.leaseId ?? null)) throw new DurableWorkError('Invalid reconciliation response.');
      return { lease: value.lease, nextAfterLeaseId: value.nextAfterLeaseId as string | null };
    });
  }
  async status(): Promise<DurableStatus> { return this.#call('status', {}, status); }
  close(): Promise<void> {
    this.#closing ??= this.#call('close', {}, value => { if (value.closed !== true) throw new DurableWorkError('Close was not confirmed.'); })
      .finally(() => { this.#child.stdin.end(); this.#child.kill(); });
    return this.#closing;
  }
  #fail(message: string): void {
    this.#fatal ??= new DurableWorkError(message);
    for (const waiter of this.#pending.values()) { clearTimeout(waiter.timer); waiter.reject(this.#fatal); }
    this.#pending.clear(); this.#buffer = ''; this.#child.kill();
  }
  #call<T>(op: string, fields: ObjectValue, parse: (value: ObjectValue) => T): Promise<T> {
    if (this.#fatal) return Promise.reject(this.#fatal);
    if (this.#closing || this.#pending.size >= 32) return Promise.reject(new DurableWorkError('Endpoint closing or 32 pending requests reached.'));
    const id = this.#nextId++;
    if (!Number.isSafeInteger(id)) return Promise.reject(new DurableWorkError('Request identity space exhausted.'));
    let frame: string;
    try { frame = JSON.stringify({ ...fields, id, protocol: 1, op }); }
    catch (error) { return Promise.reject(error); }
    if (Buffer.byteLength(frame, 'utf8') > MAX_FRAME_BYTES) return Promise.reject(new DurableWorkError('Request exceeds frame byte bound.'));
    return new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => this.#fail(`Durable '${op}' timed out; reconcile the original store before retrying.`), this.#timeout);
      timer.unref();
      this.#pending.set(id, { timer, reject, accept: value => {
        if (value.ok === false) { reject(new DurableWorkError(typeof value.error === 'string' ? value.error : 'Durable operation refused.')); return; }
        try { resolve(parse(value)); }
        catch (error) { reject(error); this.#fail('Invalid durable operation response; reconcile the original store.'); }
      } });
      this.#child.stdin.write(frame + '\n', error => { if (error) this.#fail(`Durable write failed: ${error.message}; reconcile the original store.`); });
    });
  }
}

File: bindings/python/aidotnet_evolution/durable.py
"""Version-1 local worker protocol. IDs and resource amounts stay exact decimal strings.

The host directory and pipe are trusted. A remote supervisor must supply authentication,
hardware isolation and durable physical-operation receipts. Reconciliation is not permission
to execute a lost operation again. Closing this client never refunds unknown reservations.
"""

from __future__ import annotations

import json
import math
import re
import subprocess
import threading
from datetime import datetime
from typing import Any, Mapping, Sequence

MAX_FRAME_BYTES = 16 * 1024 * 1024
_INT64_MAX = (1 << 63) - 1
_UINT64_MAX = (1 << 64) - 1
_OUTCOMES = {"completed", "failed", "rejected", "canceled"}
_DISPOSITIONS = {"accepted", "duplicate", "stale", "duplicate-stale", "unknown-lease", "budget-violation"}
_HEARTBEATS = {"renewed", "expired", "canceled", "completed", "unknown-lease"}


class DurableWorkError(RuntimeError):
    """A refused operation or an uncertain transport failure requiring reconciliation."""


def _unsigned(value: Any, maximum: int = _INT64_MAX) -> bool:
    return isinstance(value, str) and bool(re.fullmatch(r"0|[1-9][0-9]{0,19}", value)) and int(value) <= maximum


def _integer(value: Any, maximum: int = 2_147_483_647) -> bool:
    return type(value) is int and 1 <= value <= maximum


def _amounts(value: Any) -> bool:
    return isinstance(value, dict) and len(value) <= 32 and all(
        isinstance(key, str) and 0 < len(key) <= 64 and key.strip()
        and not any(ord(char) < 32 or 127 <= ord(char) <= 159 for char in key)
        and isinstance(amount, str) and re.fullmatch(r"(0|[1-9][0-9]{0,28})(\.[0-9]{0,27}[1-9])?", amount)
        for key, amount in value.items()
    )


def _identity(value: Any) -> bool:
    return isinstance(value, dict) and isinstance(value.get("runId"), str) and 0 < len(value["runId"]) <= 256 \
        and _unsigned(value.get("evaluationId")) and _integer(value.get("attempt")) \
        and isinstance(value.get("leaseId"), str) and bool(re.fullmatch(r"[0-9a-f]{32}", value["leaseId"]))


def _lease(value: Any) -> bool:
    if not isinstance(value, dict) or not _identity(value.get("identity")) or not isinstance(value.get("workerId"), str) \
            or not value["workerId"] or not isinstance(value.get("canonicalGenomeId"), str) \
            or not isinstance(value.get("payload"), str) or not _integer(value.get("deliveryNumber"), 16) \
            or not isinstance(value.get("expiresAt"), str):
        return False
    try:
        return datetime.fromisoformat(value["expiresAt"].replace("Z", "+00:00")).utcoffset() is not None
    except ValueError:
        return False


def _result(value: Any) -> bool:
    return value is None or (isinstance(value, dict) and _identity(value.get("identity"))
        and isinstance(value.get("payload"), str) and isinstance(value.get("provenance"), str)
        and _amounts(value.get("actual")) and isinstance(value.get("outcome"), str)
        and value["outcome"] in _OUTCOMES and type(value.get("accepted")) is bool)


def _require(condition: bool, message: str) -> None:
    if not condition:
        raise DurableWorkError(message)


def _unique_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise DurableWorkError("Duplicate response field: " + key)
        result[key] = value
    return result


def _json(text: str) -> Any:
    def invalid_constant(value: str) -> None:
        raise DurableWorkError("Non-JSON numeric constant: " + value)
    return json.loads(text, object_pairs_hook=_unique_object, parse_constant=invalid_constant)


def parse_evaluation_payload(text: str) -> dict[str, Any]:
    """Read the engine bridge envelope without coercing Int64 IDs or UInt64 seeds to floats."""
    _require(len(text.encode("utf-8")) <= 1024 * 1024, "Evaluation envelope exceeds one MiB.")
    value = _json(text)
    _require(isinstance(value, dict) and type(value.get("schema")) is int and value["schema"] == 1
        and isinstance(value.get("genomePayload"), str) and isinstance(value.get("canonicalGenomeId"), str)
        and bool(value["canonicalGenomeId"].strip()) and _unsigned(value.get("evaluationId"))
        and _integer(value.get("attempt")) and _unsigned(value.get("rootSeed"), _UINT64_MAX)
        and _unsigned(value.get("seedStream"), _UINT64_MAX), "Invalid versioned evaluation envelope.")
    return value


class DurableWorkClient:
    """Serialized, bounded requests to one owned host process. Use as a context manager.

    Supply the native executable as host_path, or host_path='dotnet' and the managed DLL
    in host_args. --durable is appended. On timeout the exact owned process is killed and
    the endpoint stays failed; callers must reopen the same journal and reconcile explicitly.
    """

    def __init__(self, config: Mapping[str, Any], *, host_path: str,
                 host_args: Sequence[str] = (), request_timeout: float = 120) -> None:
        _require(type(request_timeout) in (int, float) and math.isfinite(request_timeout)
            and 0 < request_timeout <= 2_147_483.647, "request_timeout must be finite and positive.")
        _require(_amounts(config.get("limits")), "Resources require canonical decimal-string amounts.")
        self._lock = threading.RLock()
        self._timeout = request_timeout
        self._next_id = 1
        self._failed: DurableWorkError | None = None
        self._closed = False
        self._stderr = b""
        self._run_id = config.get("runId")
        self._compatibility = config.get("compatibilityHash")
        self._process = subprocess.Popen([host_path, *host_args, "--durable"], stdin=subprocess.PIPE,
            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
            creationflags=getattr(subprocess, "CREATE_NO_WINDOW", 0))
        self._stderr_thread = threading.Thread(target=self._drain_stderr, daemon=True)
        self._stderr_thread.start()
        try:
            self._validate_status(self._call("open", config=dict(config)))
        except BaseException:
            self._terminate()
            raise

    @property
    def pid(self) -> int:
        """The exact child PID, for diagnostics/supervision only."""
        return self._process.pid

    def _drain_stderr(self) -> None:
        try:
            while True:
                chunk = self._process.stderr.read(4096)
                if not chunk:
                    return
                self._stderr = (self._stderr + chunk)[-4096:]
        except (OSError, ValueError):
            return

    def _terminate(self) -> None:
        if self._process.poll() is None:
            self._process.kill()
        try:
            self._process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            pass
        for stream in (self._process.stdin, self._process.stdout, self._process.stderr):
            stream.close()

    def _fatal(self, message: str) -> None:
        self._failed = self._failed or DurableWorkError(message)
        self._terminate()
        raise self._failed

    def _call(self, op: str, **fields: Any) -> dict[str, Any]:
        with self._lock:
            if self._failed:
                raise self._failed
            _require(not self._closed, "Endpoint is closed.")
            request_id = self._next_id
            _require(request_id <= (1 << 53) - 1, "Request identity space exhausted.")
            self._next_id += 1
            frame = json.dumps({**fields, "id": request_id, "protocol": 1, "op": op}, ensure_ascii=False,
                allow_nan=False, separators=(",", ":")).encode("utf-8")
            _require(len(frame) <= MAX_FRAME_BYTES, "Request exceeds frame byte bound.")
            completed: list[bytes | BaseException] = []

            def exchange() -> None:
                try:
                    remaining = memoryview(frame + b"\n")
                    while remaining:
                        written = self._process.stdin.write(remaining)
                        if not written:
                            raise BrokenPipeError("Host input pipe closed.")
                        remaining = remaining[written:]
                    self._process.stdin.flush()
                    completed.append(self._process.stdout.readline(MAX_FRAME_BYTES + 2))
                except BaseException as error:
                    completed.append(error)

            # A stalled write is bounded by the same deadline as a stalled read. Only one
            # exchange can exist at a time; killing this owned child releases either pipe.
            thread = threading.Thread(target=exchange, daemon=True)
            thread.start()
            thread.join(self._timeout)
            if thread.is_alive():
                self._fatal(f"Durable '{op}' timed out; reconcile the original store before retrying.")
            if not completed or isinstance(completed[0], BaseException):
                self._fatal("Durable pipe failed; reconcile unacknowledged work. " + str(completed))
            raw = completed[0]
            if len(raw) > MAX_FRAME_BYTES or not raw.endswith(b"\n"):
                self._fatal("Oversized, truncated or closed durable response; reconcile the original store.")
            try:
                value = _json(raw.decode("utf-8"))
                _require(isinstance(value, dict) and type(value.get("id")) is int and value["id"] == request_id
                    and type(value.get("protocol")) is int and value["protocol"] == 1 and type(value.get("ok")) is bool,
                    "Invalid correlation or durable protocol downgrade.")
            except (ValueError, DurableWorkError) as error:
                self._fatal(str(error))
            if not value["ok"]:
                _require(isinstance(value.get("error"), str), "Malformed refusal from durable host.")
                raise DurableWorkError(value["error"])
            return value

    def _checked(self, value: Any, condition: bool, message: str) -> Any:
        if not condition:
            self._fatal(message + " Reconcile the original store.")
        return value

    def _validate_status(self, value: dict[str, Any]) -> dict[str, Any]:
        source = value.get("sourceSessionId")
        return self._checked(value, value.get("runId") == self._run_id and value.get("compatibilityHash") == self._compatibility
            and type(value.get("wasRecovered")) is bool and "sourceSessionId" in value
            and (source is None or (isinstance(source, str) and bool(re.fullmatch(r"[0-9a-f]{32}", source))))
            and value.get("supportsExactSearchContinuation") is False and value.get("searchState") == "not-owned"
            and isinstance(value.get("searchContinuationGuarantee"), str) and "fork" in value["searchContinuationGuarantee"]
            and _amounts(value.get("spent")) and _amounts(value.get("reserved"))
            and all(_unsigned(value.get(key)) for key in ("admitted", "settled", "denied"))
            and type(value.get("maximumViolated")) is bool, "Invalid durable status response.")

    def enqueue(self, job: Mapping[str, Any]) -> bool:
        _require(_unsigned(job.get("evaluationId")) and _integer(job.get("attempt")), "Invalid logical work identity.")
        _require(_amounts(job.get("estimated")) and _amounts(job.get("maximum")), "Resources require decimal-string amounts.")
        value = self._call("enqueue", job=dict(job)).get("enqueued")
        return self._checked(value, type(value) is bool, "Invalid enqueue response.")

    def claim(self, worker: Mapping[str, Any]) -> dict[str, Any] | None:
        """Null means no compatible work now, never search completion."""
        value = self._call("claim", worker=dict(worker))
        lease = value.get("lease")
        valid = (value.get("available") is False and "lease" in value and lease is None) or (
            value.get("available") is True and _lease(lease) and lease["workerId"] == worker.get("workerId")
            and lease["identity"]["runId"] == self._run_id)
        return self._checked(lease, valid, "Invalid claim response.")

    def heartbeat(self, identity: Mapping[str, Any], worker_id: str) -> str:
        _require(_identity(identity), "The original complete durable identity is required.")
        value = self._call("heartbeat", identity=dict(identity), workerId=worker_id).get("status")
        return self._checked(value, isinstance(value, str) and value in _HEARTBEATS, "Invalid heartbeat response.")

    def cancel(self, evaluation_id: str, attempt: int) -> bool:
        _require(_unsigned(evaluation_id) and _integer(attempt), "Invalid logical work identity.")
        value = self._call("cancel", evaluationId=evaluation_id, attempt=attempt).get("canceled")
        return self._checked(value, type(value) is bool, "Invalid cancel response.")

    def commit(self, receipt: Mapping[str, Any]) -> str:
        _require(_identity(receipt.get("identity")) and _amounts(receipt.get("actual"))
            and isinstance(receipt.get("outcome"), str) and receipt["outcome"] in _OUTCOMES, "A complete original ticket and actual decimal receipt are required.")
        value = self._call("commit", **dict(receipt)).get("disposition")
        return self._checked(value, isinstance(value, str) and value in _DISPOSITIONS, "Invalid commit response.")

    def result(self, evaluation_id: str, attempt: int) -> dict[str, Any] | None:
        _require(_unsigned(evaluation_id) and _integer(attempt), "Invalid logical work identity.")
        reply = self._call("result", evaluationId=evaluation_id, attempt=attempt)
        value = reply.get("result")
        return self._checked(value, "result" in reply and _result(value) and (value is None or (
            value["identity"]["evaluationId"] == evaluation_id and value["identity"]["attempt"] == attempt
            and value["identity"]["runId"] == self._run_id)), "Invalid logical result response.")

    def delivery(self, identity: Mapping[str, Any], worker_id: str) -> dict[str, Any] | None:
        _require(_identity(identity), "The original complete durable identity is required.")
        reply = self._call("delivery", identity=dict(identity), workerId=worker_id)
        value = reply.get("result")
        return self._checked(value, "result" in reply and _result(value) and (value is None or value["identity"] == identity), "Invalid delivery receipt.")

    def unsettled(self, worker_id: str, after_lease_id: str | None = None) -> dict[str, Any]:
        """One keyset page for reconciliation only; new concurrent claims require restarting a scan."""
        fields = {"workerId": worker_id}
        if after_lease_id is not None:
            fields["afterLeaseId"] = after_lease_id
        value = self._call("unsettled", **fields)
        lease = value.get("lease")
        valid = value.get("reconciliationOnly") is True and "lease" in value and "nextAfterLeaseId" in value and (
            (lease is None and value["nextAfterLeaseId"] is None) or (_lease(lease)
                and lease["workerId"] == worker_id and lease["identity"]["runId"] == self._run_id
                and value["nextAfterLeaseId"] == lease["identity"]["leaseId"]))
        return self._checked(value, valid, "Invalid reconciliation response.")

    def status(self) -> dict[str, Any]:
        return self._validate_status(self._call("status"))

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            try:
                value = self._call("close")
                _require(value.get("closed") is True, "Close was not confirmed.")
            finally:
                self._closed = True
                self._terminate()

    def __enter__(self) -> DurableWorkClient:
        return self

    def __exit__(self, kind: Any, value: Any, traceback: Any) -> None:
        if kind is None:
            self.close()
        else:
            try:
                self.close()
            except Exception:
                pass  # Preserve the evaluator's original failure, but always tear down.
